src/v/pandaproxy/schema_registry/seq_writer.cc

```diff
@@ -220,4 +227,186 @@ ss::future<bool> seq_writer::write_config(
+          .seq{write_at}, .node{my_node_id}, .sub{sub}, .version{version}};
+        vlog(plog.debug, "seq_writer::delete_subject_version {}", key);
+        auto value = schema_value{
+          .sub{sub},
+          .version{version},
+          .type = ss.type,
+          .id{ss.id},
+          .schema{std::move(ss.definition)},
+          .deleted{is_deleted::yes}};
+
+        auto batch = as_record_batch(key, value);
+
+        auto success = co_await seq.produce_and_check(
+          write_at, std::move(batch));
+        if (success) {
+            auto applier = consume_to_store(seq._store, seq);
+            co_await applier.apply(write_at, key, value);
+            seq.advance_offset_inner(write_at);
+            co_return true;
+        } else {
+            // Pass up a None, our caller's cue to retry
+            co_return std::nullopt;
+        }
+    };
+
+    co_return co_await sequenced_write(do_write);
+}
+
+ss::future<std::vector<schema_version>>
+seq_writer::delete_subject_impermanent(subject sub) {
+    vlog(plog.debug, "delete_subject_impermanent sub={}", sub);
+    auto do_write = [sub](model::offset write_at, seq_writer& seq)
+      -> ss::future<std::optional<std::vector<schema_version>>> {
+        // Grab the versions before they're gone.
+        std::vector<schema_version> versions = co_await seq._store.get_versions(
+          sub, include_deleted::yes);
+
+        // Inspect the subject to see if its already deleted
+        if (co_await seq._store.is_subject_deleted(sub)) {
+            co_return std::make_optional(versions);
+        }
+
+        // Proceed to write
+        auto my_node_id = config::shard_local_cfg().node_id();
+        auto version = versions.back();
+        auto key = delete_subject_key{
+          .seq{write_at}, .node{my_node_id}, .sub{sub}};
+        auto value = delete_subject_value{.sub{sub}, .version{version}};
+        auto batch = as_record_batch(key, value);
+
+        auto success = co_await seq.produce_and_check(
+          write_at, std::move(batch));
+        if (success) {
+            auto applier = consume_to_store(seq._store, seq);
+            co_await applier.apply(write_at, key, value);
+            seq.advance_offset_inner(write_at);
+            co_return versions;
+        } else {
+            // Pass up a None, our caller's cue to retry
+            co_return std::nullopt;
+        }
+    };
+
+    co_return co_await sequenced_write(do_write);
+}
+
+/// Permanent deletions (i.e. writing tombstones for previous sequenced
+/// records) do not themselves need sequence numbers.
+/// Include a version if we are only to hard delete that version, otherwise
+/// will hard-delete the whole subject.
+ss::future<std::vector<schema_version>> seq_writer::delete_subject_permanent(
+  subject sub, std::optional<schema_version> version) {
+    return container().invoke_on(0, _smp_opts, [sub, version](seq_writer& seq) {
+        auto units = ss::get_units(seq._write_sem, 1);
```

Comment:
Do `units` need a lifetime extension to cover the operation?

Reply:
Fixed with `with_semaphore`